server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -241,9 +268,9 @@ private HitContext prepareHitContext(SearchContext context,
                                          CheckedBiConsumer<Integer, FieldsVisitor, IOException> storedFieldReader) throws IOException {
         if (nestedDocuments.advance(docId - subReaderContext.docBase) == null) {
             return prepareNonNestedHitContext(
-                context, fieldsVisitor, docId, storedToRequestedFields, subReaderContext, storedFieldReader);
+                context, profiler, fieldsVisitor, docId, storedToRequestedFields, subReaderContext, storedFieldReader);
         } else {
-            return prepareNestedHitContext(context, docId, nestedDocuments, storedToRequestedFields,
+            return prepareNestedHitContext(context, profiler, docId, nestedDocuments, storedToRequestedFields,
```

Comment:
It would be nice to have a test or example of how the output looks when there are nested hits.

Reply:
++

Reply:
I've added support for nested and parent/child based inner_hits - you see the whole sub-sub-fetch phase as a child of the phase. It looks like:
```
  1>             "fetch" : {
  1>               "type" : "fetch",
  1>               "description" : "fetch",
  1>               "time_in_nanos" : 4181937,
  1>               "breakdown" : {
  1>                 "load_stored_fields" : 133903,
  1>                 "load_stored_fields_count" : 1,
  1>                 "next_reader" : 13716,
  1>                 "next_reader_count" : 1
  1>               },
  1>               "debug" : {
  1>                 "stored_fields" : [
  1>                   "_id",
  1>                   "_routing",
  1>                   "_source"
  1>                 ]
  1>               },
  1>               "children" : [
  1>                 {
  1>                   "type" : "inner_hits",
  1>                   "description" : "fetches matching nested or parent or child documents",
  1>                   "time_in_nanos" : 3046811,
  1>                   "breakdown" : {
  1>                     "process_count" : 1,
  1>                     "process" : 3046193,
  1>                     "next_reader" : 618,
  1>                     "next_reader_count" : 1
  1>                   },
  1>                   "children" : [
  1>                     {
  1>                       "type" : "question",
  1>                       "description" : "inner_hits",
  1>                       "time_in_nanos" : 101738,
  1>                       "breakdown" : {
  1>                         "load_stored_fields" : 23394,
  1>                         "load_stored_fields_count" : 1,
  1>                         "next_reader" : 2734,
  1>                         "next_reader_count" : 1
  1>                       },
  1>                       "debug" : {
  1>                         "stored_fields" : [
  1>                           "_id",
  1>                           "_routing",
  1>                           "_source"
  1>                         ]
  1>                       },
  1>                       "children" : [
  1>                         {
  1>                           "type" : "source",
  1>                           "description" : "load _source",
  1>                           "time_in_nanos" : 4243,
  1>                           "breakdown" : {
  1>                             "process_count" : 1,
  1>                             "process" : 3949,
  1>                             "next_reader" : 294,
  1>                             "next_reader_count" : 1
  1>                           },
  1>                           "debug" : {
  1>                             "fast_path" : 1,
  1>                             "loaded_nested" : 0
  1>                           }
  1>                         }
  1>                       ]
  1>                     }
  1>                   ]
  1>                 },

```

It just exposes whatever is in the sub-phase without thinking about what the sub-phase tries to share. Hopefully that should all come out in the timings.

Reply:
I've reverted this bit so we can do it in a follow up.